src/citrine/resources/data_concepts.py

```diff
@@ -273,15 +274,21 @@ def _build_list_of_soft_links(obj, obj_with_soft_links, field: str, reverse_fiel
         if linked_objects is None:
             return
 
+        # Cycle through linked objects in obj_with_soft_link and if they are not LinkByUID,
+        # build them and then set their `reverse_field` field to dc_obj
         for linked_obj in linked_objects:
-            # Cycle through linked objects and if they are not LinkByUID, build them and then
-            # set their `reverse_field` field to obj
             assert isinstance(linked_obj, DictSerializable)
             if isinstance(linked_obj, LinkByUID):
                 pass
+            # Sever the link between linked_obj and obj_with_soft_links.
+            # This prevents infinite loops in the next step.
             setattr(linked_obj, reverse_field, None)
-            meas_object = linked_type.build(linked_obj, session)
-            setattr(meas_object, reverse_field, obj)
+            # Build a DataConcepts instance of linked_obj
+            dc_linked_obj = linked_type.build(linked_obj, session)
+            # Establish the link between the DataConcepts instances of obj and linked_obj
+            setattr(dc_linked_obj, reverse_field, dc_obj)
+            # Re-establish the link between linked_obj and obj_with_soft_links
+            setattr(linked_obj, reverse_field, obj_with_soft_links)
```

Comment:
This is the only new line of functional code. The other changes are all fleshed-out documentation and changing the names of some variables to be more descriptive.